x/evm/keeper/vote_handler.go

```diff
@@ -0,0 +1,175 @@
+package keeper
+
+import (
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/axelarnetwork/utils/slices"
+
+	"github.com/axelarnetwork/axelar-core/x/evm/types"
+	nexus "github.com/axelarnetwork/axelar-core/x/nexus/exported"
+	vote "github.com/axelarnetwork/axelar-core/x/vote/exported"
+)
+
+// NewVoteHandler returns the handler for processing vote delivered by the vote module
+func NewVoteHandler(cdc codec.Codec, keeper types.BaseKeeper, nexus types.Nexus) vote.VoteHandler {
+	return func(ctx sdk.Context, result *vote.Vote) error {
+		events, err := types.UnpackEvents(cdc, result.Results)
+		if err != nil {
+			return err
+		}
+
+		if len(events) == 0 {
+			return fmt.Errorf("vote result has no events")
+		}
+
+		chainName := events[0].Chain
+		if slices.Any(events, func(event types.Event) bool { return event.Chain != chainName }) {
+			return fmt.Errorf("events are not from the same source chain")
+		}
+
+		chain, ok := nexus.GetChain(ctx, chainName)
+		if !ok {
+			return fmt.Errorf("%s is not a registered chain", chainName)
+		}
+		if !keeper.HasChain(ctx, chainName) {
+			return fmt.Errorf("%s is not an evm chain", chainName)
+		}
+
+		chainK := keeper.ForChain(chain.Name)
+		cacheCtx, writeCache := ctx.CacheContext()
+
+		err = handleEvents(cacheCtx, chainK, nexus, events, chain)
+		if err != nil {
+			// set events to failed, we will deal with later
+			for _, e := range events {
+				chainK.FailEvent(ctx, e)
```

Comment:
Not checking error returned here?

Reply:
if the event is already completed/failed, it will return error. No need to check this in my opinion, wdyt?

Reply:
I see. I would then panic since it's impossible.